content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+<!-- prettier-ignore -->
+```js
+// The app shows a search box and a list of names. The list is
+// controlled by the searchValue state variable, which is updated
+// by the search box.
+function App() {
+  const [searchValue, setSearchValue] = React.useState();
+
+  function handleChange(value) {
+    setSearchValue(value);
+  }
+
+  return (
+    <div>
+      <SearchBox onChange={handleChange} />
+      <NameList searchValue={searchValue} />
+    </div>
+  );
+}
+
+// The search box renders a native HTML input element and keeps
+// it controlled using the inputValue variable. When a new key
+// is pressed, it will first update the local inputValue, then
+// it will update the App component’s searchValue, and then it will
+// simulate an analytics notification to our servers.
+function SearchBox(props) {
+  const [inputValue, setInputValue] = React.useState();
+
+  function handleChange(event) {
+    const value = event.target.value;
+
+    setInputValue(value);
+    props.onChange(value);
+    sendAnalyticsNotification(value);
+  };
+
+  return (
+    <input
+      type="text"
+      value={inputValue}
+      onChange={handleChange}
+    />
+  );
+}
+
+ReactDOM.render(<App />, container);
+```
+
+_ℹ️ This example uses [React Hooks](https://reactjs.org/docs/hooks-intro.html). If you’re not familiar with this new React feature, take a look at the [CodeSandbox code](https://codesandbox.io/s/j3zrqpzkr5). Additionally, you might wonder why we use two different state variables for this example. We’ll find out why this is needed in a bit._
+
+Try it out! Type a name (e.g. “Ada Stewart”) in the search box below and see how it works:
+
+<iframe src="https://codesandbox.io/embed/j3zrqpzkr5?fontsize=14&hidenavigation=1&view=preview" style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+You might notice that the interface is not very responsive. To amplify the issue, I artificially slowed down the rendering time of the names list. And since this list is big, it has a significant impact on the application’s performance. This is not good 😰.
+
+Our users expect immediate feedback, but the app is unresponsive for seconds after a keystroke. To understand what’s going on, let’s take a look at the DevTools’ Performance tab. Here’s a screenshot of a recording I made while I type the name “Ada” into the search box:
+
+![Screenshot of Chrome DevTools that shows that the three keypress events take 733ms to render.](devtools-sync.png)
+
+We can see there are a lot of red triangles, which is usually not a good sign. For every keystroke, we see a `keypress` event being fired. All three events run within one frame,[^4] which causes the frame to take **733ms**. That’s way above our average frame budget of 16ms.
```

Comment:
What does it take 733ms to do?

"which causes the frame to take **733ms**."

Reply:
> which causes the frame to extend to 733ms.

One frame is kinda a time period which should be as small as possible (so that we can maintain a 16ms frame rate). I think we can say that the frame is extended to multiple milliseconds.

Reply:
Which is better:

"All three events run within one frame,[^4] which causes the frame to extend to **733ms**."

"All three events run within one frame,[^4] which causes the length of the frame to extend to **733ms**."

"All three events run within one frame,[^4] which causes the duration of the frame to extend to **733ms**."

I'm leaning toward the last one, but not sure if that's correct.

Reply:
Yeah that one is the best :-) Fixed it!